{% extends "layout.html" %}

{% block title %}
<title>{{ config['BROWSER_NAME'] }} - {{ transcript.transcript_id }}</title>
{% endblock %}

{% block body %}
    <!-- Render context vars in JS here -->
    <script type="text/javascript">
        window.transcript = {{ transcript|tojson(separators=(',',':'))|safe }};
        window.table_variants = {{ variants_in_transcript|tojson(separators=(',',':'))|safe }};
        window.variants_in_transcript = window.table_variants;
        window.coverage_stats = {{ coverage_stats|tojson(separators=(',',':'))|safe }};
        window.gene = {{ gene|tojson(separators=(',',':'))|safe }};
        window.page_name = window.gene.symbol + '_' + window.transcript.transcript_id;


        $(document).ready(function() {
            $('#consequence_other_variant_button')
            .addClass('disabled')
            .attr('disabled', 'disabled')
            .attr('data-tooltip', "Loading...");
        });

        $.when($.getJSON("/api/variants_in_transcript/" + window.transcript.transcript_id),
               $.Deferred(function() { $(this.resolve); })) //document.ready
        .done(function(transcript_data) {
            // Somehow these arguments get wrapped up, so we want transcript_data[0].

            $('#consequence_other_variant_button')
            .attr('data-tooltip', "Show all variants")
            .removeClass('disabled')
            .removeAttr('disabled')
            .one('change', function() {
                with_waiting_notice(function() {
                    //add variants to the variant_table
                    window.table_variants = transcript_data[0];
                    build_variant_table()

                    //add variants to the variant graph
                    window.variants_in_transcript = transcript_data[0];
                    window.precalc_coding_coordinates(window.variants_in_transcript);
                    add_variants_to_variants_plot();
                    update_variants();
                });
            });
        })
        .fail(function() {
            console.log("XHR failed for transcript.");
        });

    </script>
    <div class="container-fluid">
        <div class="row">
            <div class="col-xs-11 col-xs-offset-1">
                <h1>{{ gene.gene_name }}</h1>
                <h3>Transcript: {{ transcript.transcript_id }} (<span id="number_of_exons"></span> coding exons)</h3>
                <script>
                  var coding_exons = _.filter(window.transcript.exons, function(d) {
                      return d.feature_type == 'CDS';
                  });
                  $('#number_of_exons').html(coding_exons.length);
                </script>
            </div>
        </div>
        <div class="row">
            <div class="col-xs-11 col-xs-offset-1">
                <dl class="dl-horizontal">
                    <dt>Number of variants</dt>
                    <dd>{{ num_variants_in_transcript }}</dd>
                    <dt class="hidden-xs">UCSC Browser</dt>
                    <dd class="hidden-xs">
                        <a href="http://genome.ucsc.edu/cgi-bin/hgTracks?db=hg19&position=chr{{ transcript.chrom }}%3A{{ transcript.start - 1 }}-{{ transcript.stop - 1 }}" target="_blank">
                            {{ transcript.chrom }}:{{ transcript.start - 1 }}-{{ transcript.stop - 1 }}
                            <i class="fa fa-external-link"></i>
                        </a>
                    </dd>
                </dl>
                <div class="dropdown">
                    <button class="btn btn-default dropdown-toggle" type="button" id="transcript_dropdown" data-toggle="dropdown">
                        Other transcripts in this gene
                        <span class="caret"></span>
                    </button>
                    <ul class="dropdown-menu" role="menu" aria-labelledby="transcript_dropdown">
                        {% for transcript in gene.transcripts %}
                            <li role="presentation">
                                <a role="menuitem" tabindex="-1" href="/transcript/{{ transcript.transcript_id }}">
                                    {{ transcript.transcript_id }}
                                    {% if transcript.transcript_id == gene.canonical_transcript %}
                                        *
                                    {% endif %}
                                </a>
                            </li>
                        {% endfor %}
                    </ul>
                </div>
            </div>
        </div>
        <div class="row">
            <div class="col-xs-12 col-md-10 col-md-offset-1">
                {% with plot_id = transcript.transcript_id %}
                    {% include 'transcript_display.html' %}
                {% endwith %}
            </div>
         </div>
         <div class="row">
            <div class="col-xs-12 col-md-10 col-md-offset-1">
                {% if variants_in_transcript %}
                    {% with chrom = variants_in_transcript[0].chrom %}
                        {% include 'variant_table.html' %}
                    {% endwith %}
                {% else %}
                    No variants in this gene.
                {% endif %}
            </div>
        </div>
    </div>
{% endblock %}